// @healthsync/core/dist/esm/index.js
/**
 * HealthSync SDK - Core Package
 *
 * Universal health data integration platform with plugin architecture.
 * Provides unified access to multiple health data sources through a single, elegant API.
 *
 * @module @healthsync/core
 * @version 1.0.0
 */
// ============================================================================
// Unified Data Models
// ============================================================================
export { 
// Enums
HealthSource, DataType, DataQuality, HeartRateContext, SleepStage, ActivityType, 
// Type Guards & Utilities
isDataType, isValidHealthData, } from './models/unified-data';
// ============================================================================
// Plugin System
// ============================================================================
export { 
// Enums
ConnectionStatus, ErrorAction, 
// Base Class
BasePlugin, } from './plugins/plugin-interface';
// ============================================================================
// Configuration & Types
// ============================================================================
export { 
// Enums
LogLevel, CacheLayer, 
// Error Classes
SDKError, ConfigurationError, AuthenticationError, ConnectionError, DataFetchError, RateLimitError, ValidationError, PluginError, CacheError, SyncError, NetworkError, 
// Default Configurations
DEFAULT_SDK_CONFIG, DEFAULT_SYNC_OPTIONS, } from './types/config';
// ============================================================================
// Query & Response Types
// ============================================================================
export { 
// Enums
SDKEvent, } from './types/query';
// ============================================================================
// Main SDK
// ============================================================================
export { HealthSyncSDK } from './sdk';
// ============================================================================
// Plugin Registry
// ============================================================================
export { PluginRegistry } from './plugins/plugin-registry';
// ============================================================================
// Health Connect Plugin
// ============================================================================
export { HealthConnectPlugin, HealthConnectPermission, HealthConnectRecordType, HealthConnectSleepStage, HealthConnectExerciseType, HealthConnectAvailability, HEALTH_CONNECT_TYPE_MAP, DEFAULT_HEALTH_CONNECT_CONFIG, } from './plugins/health-connect';
// ============================================================================
// Data Normalizer
// ============================================================================
export { DataNormalizer, } from './normalizer/data-normalizer';
export { UnitConverter, } from './normalizer/unit-converter';
export { DataValidator, } from './normalizer/validator';
export { QualityScorer, } from './normalizer/quality-scorer';
// ============================================================================
// Cache System
// ============================================================================
export { generateCacheKey, parseCacheKey, } from './cache/cache-provider';
export { MemoryCacheProvider } from './cache/memory-cache';
export { CacheManager, } from './cache/cache-manager';
// ============================================================================
// Utilities
// ============================================================================
export { EventEmitter, } from './utils/event-emitter';
// ============================================================================
// Version
// ============================================================================
/**
 * Current SDK version
 */
export const VERSION = '1.0.0';
/**
 * SDK name
 */
export const SDK_NAME = 'HealthSync SDK';
/**
 * SDK description
 */
export const SDK_DESCRIPTION = 'Universal health data integration platform';
