// @healthsync/core/dist/models/unified-data.js
"use strict";
/**
 * Unified Health Data Models
 *
 * This module defines the standardized data structures used across all health data sources.
 * All plugins must transform their native data formats into these unified models.
 *
 * @module models/unified-data
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityType = exports.SleepStage = exports.HeartRateContext = exports.DataQuality = exports.DataType = exports.HealthSource = void 0;
exports.isDataType = isDataType;
exports.isValidHealthData = isValidHealthData;
/**
 * Enumeration of supported health data sources
 *
 * @enum {string}
 */
var HealthSource;
(function (HealthSource) {
    /** Google Health Connect (Android) */
    HealthSource["HEALTH_CONNECT"] = "health_connect";
    /** Apple HealthKit (iOS) */
    HealthSource["APPLE_HEALTH"] = "apple_health";
    /** Fitbit Cloud API */
    HealthSource["FITBIT"] = "fitbit";
    /** Garmin Connect API */
    HealthSource["GARMIN"] = "garmin";
    /** Oura Ring API */
    HealthSource["OURA"] = "oura";
    /** Whoop API */
    HealthSource["WHOOP"] = "whoop";
    /** Strava API */
    HealthSource["STRAVA"] = "strava";
    /** MyFitnessPal API */
    HealthSource["MYFITNESSPAL"] = "myfitnesspal";
    /** Generic/Unknown source */
    HealthSource["UNKNOWN"] = "unknown";
})(HealthSource || (exports.HealthSource = HealthSource = {}));
/**
 * Enumeration of supported health data types
 *
 * @enum {string}
 */
var DataType;
(function (DataType) {
    /** Step count data */
    DataType["STEPS"] = "steps";
    /** Heart rate measurements (BPM) */
    DataType["HEART_RATE"] = "heart_rate";
    /** Sleep session data */
    DataType["SLEEP"] = "sleep";
    /** Physical activity/workout sessions */
    DataType["ACTIVITY"] = "activity";
    /** Calories burned */
    DataType["CALORIES"] = "calories";
    /** Distance traveled */
    DataType["DISTANCE"] = "distance";
    /** Active minutes/energy */
    DataType["ACTIVE_MINUTES"] = "active_minutes";
    /** Blood oxygen saturation (SpO2) */
    DataType["BLOOD_OXYGEN"] = "blood_oxygen";
    /** Blood pressure measurements */
    DataType["BLOOD_PRESSURE"] = "blood_pressure";
    /** Blood glucose levels */
    DataType["BLOOD_GLUCOSE"] = "blood_glucose";
    /** Body temperature */
    DataType["BODY_TEMPERATURE"] = "body_temperature";
    /** Body weight measurements */
    DataType["WEIGHT"] = "weight";
    /** Height measurements */
    DataType["HEIGHT"] = "height";
    /** Body Mass Index */
    DataType["BMI"] = "bmi";
    /** Body fat percentage */
    DataType["BODY_FAT"] = "body_fat";
    /** Hydration/water intake */
    DataType["HYDRATION"] = "hydration";
    /** Nutrition/food intake */
    DataType["NUTRITION"] = "nutrition";
    /** Respiratory rate */
    DataType["RESPIRATORY_RATE"] = "respiratory_rate";
    /** Heart rate variability */
    DataType["HEART_RATE_VARIABILITY"] = "heart_rate_variability";
    /** Resting heart rate */
    DataType["RESTING_HEART_RATE"] = "resting_heart_rate";
    /** VO2 Max (cardio fitness) */
    DataType["VO2_MAX"] = "vo2_max";
})(DataType || (exports.DataType = DataType = {}));
/**
 * Data quality/reliability scoring
 *
 * @enum {string}
 */
var DataQuality;
(function (DataQuality) {
    /** High confidence data from medical-grade devices */
    DataQuality["HIGH"] = "high";
    /** Medium confidence data from consumer devices */
    DataQuality["MEDIUM"] = "medium";
    /** Low confidence data from estimated/derived sources */
    DataQuality["LOW"] = "low";
    /** Unknown quality */
    DataQuality["UNKNOWN"] = "unknown";
})(DataQuality || (exports.DataQuality = DataQuality = {}));
/**
 * Heart rate context/measurement type
 *
 * @enum {string}
 */
var HeartRateContext;
(function (HeartRateContext) {
    /** Resting heart rate */
    HeartRateContext["RESTING"] = "resting";
    /** Active/exercise heart rate */
    HeartRateContext["ACTIVE"] = "active";
    /** Peak heart rate during intense activity */
    HeartRateContext["PEAK"] = "peak";
    /** General/unspecified context */
    HeartRateContext["GENERAL"] = "general";
})(HeartRateContext || (exports.HeartRateContext = HeartRateContext = {}));
/**
 * Sleep stage types
 *
 * @enum {string}
 */
var SleepStage;
(function (SleepStage) {
    /** Awake during sleep session */
    SleepStage["AWAKE"] = "awake";
    /** Light sleep */
    SleepStage["LIGHT"] = "light";
    /** Deep sleep */
    SleepStage["DEEP"] = "deep";
    /** REM (Rapid Eye Movement) sleep */
    SleepStage["REM"] = "rem";
    /** Unknown or unclassified stage */
    SleepStage["UNKNOWN"] = "unknown";
})(SleepStage || (exports.SleepStage = SleepStage = {}));
/**
 * Activity/Exercise type
 *
 * @enum {string}
 */
var ActivityType;
(function (ActivityType) {
    /** Walking */
    ActivityType["WALKING"] = "walking";
    /** Running */
    ActivityType["RUNNING"] = "running";
    /** Cycling */
    ActivityType["CYCLING"] = "cycling";
    /** Swimming */
    ActivityType["SWIMMING"] = "swimming";
    /** Hiking */
    ActivityType["HIKING"] = "hiking";
    /** Yoga */
    ActivityType["YOGA"] = "yoga";
    /** Weight training */
    ActivityType["WEIGHT_TRAINING"] = "weight_training";
    /** General gym workout */
    ActivityType["WORKOUT"] = "workout";
    /** Sports activity */
    ActivityType["SPORTS"] = "sports";
    /** Dancing */
    ActivityType["DANCING"] = "dancing";
    /** Meditation */
    ActivityType["MEDITATION"] = "meditation";
    /** Other/unspecified activity */
    ActivityType["OTHER"] = "other";
})(ActivityType || (exports.ActivityType = ActivityType = {}));
/**
 * Type guard to check if data is of a specific type
 *
 * @template T - The specific health data type
 * @param {UnifiedHealthData} data - The data to check
 * @param {DataType} type - The expected data type
 * @returns {data is T} True if data matches the type
 */
function isDataType(data, type) {
    return data.dataType === type;
}
/**
 * Helper function to validate health data structure
 *
 * @param {unknown} data - Data to validate
 * @returns {data is UnifiedHealthData} True if data has valid structure
 */
function isValidHealthData(data) {
    if (!data || typeof data !== 'object') {
        return false;
    }
    const obj = data;
    return (typeof obj['id'] === 'string' &&
        typeof obj['source'] === 'string' &&
        typeof obj['dataType'] === 'string' &&
        typeof obj['timestamp'] === 'string' &&
        typeof obj['metadata'] === 'object' &&
        obj['metadata'] !== null);
}
