// @healthsync/core/dist/esm/plugins/plugin-interface.js
/**
 * Plugin Interface
 *
 * This module defines the contract that all health data plugins must implement.
 * Plugins are the bridge between the HealthSync SDK and specific health data sources.
 *
 * @module plugins/plugin-interface
 */
/**
 * Connection status enumeration
 *
 * @enum {string}
 */
export var ConnectionStatus;
(function (ConnectionStatus) {
    /** Plugin is connected and ready */
    ConnectionStatus["CONNECTED"] = "connected";
    /** Plugin is disconnected */
    ConnectionStatus["DISCONNECTED"] = "disconnected";
    /** Plugin is connecting */
    ConnectionStatus["CONNECTING"] = "connecting";
    /** Plugin is disconnecting */
    ConnectionStatus["DISCONNECTING"] = "disconnecting";
    /** Connection is in error state */
    ConnectionStatus["ERROR"] = "error";
    /** Connection requires re-authentication */
    ConnectionStatus["REQUIRES_AUTH"] = "requires_auth";
})(ConnectionStatus || (ConnectionStatus = {}));
/**
 * Error action to take when an error occurs
 *
 * @enum {string}
 */
export var ErrorAction;
(function (ErrorAction) {
    /** Retry the operation with exponential backoff */
    ErrorAction["RETRY"] = "retry";
    /** Fail immediately, do not retry */
    ErrorAction["FAIL"] = "fail";
    /** Queue the operation for later */
    ErrorAction["QUEUE"] = "queue";
    /** Request user re-authentication */
    ErrorAction["REAUTH"] = "reauth";
    /** Ignore the error and continue */
    ErrorAction["IGNORE"] = "ignore";
})(ErrorAction || (ErrorAction = {}));
/**
 * Abstract base class providing common plugin functionality
 *
 * Plugin developers can extend this class to inherit common patterns
 * like event emitting, state management, and error handling.
 *
 * @abstract
 * @class BasePlugin
 * @implements {IHealthDataPlugin}
 */
export class BasePlugin {
    constructor() {
        Object.defineProperty(this, "config", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {}
        });
        Object.defineProperty(this, "connectionStatus", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ConnectionStatus.DISCONNECTED
        });
    }
    /**
     * Check if plugin is connected
     */
    async isConnected() {
        return this.connectionStatus === ConnectionStatus.CONNECTED;
    }
    /**
     * Get connection status
     */
    async getConnectionStatus() {
        return this.connectionStatus;
    }
    /**
     * Default error handling strategy
     */
    handleError(error) {
        const errorMessage = error.message.toLowerCase();
        // Authentication errors
        if (errorMessage.includes('unauthorized') ||
            errorMessage.includes('auth') ||
            errorMessage.includes('401')) {
            return ErrorAction.REAUTH;
        }
        // Rate limit errors
        if (errorMessage.includes('rate limit') || errorMessage.includes('429')) {
            return ErrorAction.QUEUE;
        }
        // Transient network errors
        if (errorMessage.includes('timeout') ||
            errorMessage.includes('network') ||
            errorMessage.includes('503')) {
            return ErrorAction.RETRY;
        }
        // Permanent errors
        if (errorMessage.includes('400') ||
            errorMessage.includes('404') ||
            errorMessage.includes('permission denied')) {
            return ErrorAction.FAIL;
        }
        // Default to retry for unknown errors
        return ErrorAction.RETRY;
    }
    /**
     * Get plugin info
     */
    async getInfo() {
        return {
            id: this.id,
            name: this.name,
            version: this.version,
            supportedDataTypes: this.supportedDataTypes,
            requiresAuthentication: this.requiresAuthentication,
            isCloudBased: this.isCloudBased,
        };
    }
}
