// @healthsync/core/dist/index.js
"use strict";
/**
 * HealthSync SDK - Core Package
 *
 * Universal health data integration platform with plugin architecture.
 * Provides unified access to multiple health data sources through a single, elegant API.
 *
 * @module @healthsync/core
 * @version 1.0.0
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDK_DESCRIPTION = exports.SDK_NAME = exports.VERSION = exports.EventEmitter = exports.CacheManager = exports.MemoryCacheProvider = exports.parseCacheKey = exports.generateCacheKey = exports.QualityScorer = exports.DataValidator = exports.UnitConverter = exports.DataNormalizer = exports.DEFAULT_HEALTH_CONNECT_CONFIG = exports.HEALTH_CONNECT_TYPE_MAP = exports.HealthConnectAvailability = exports.HealthConnectExerciseType = exports.HealthConnectSleepStage = exports.HealthConnectRecordType = exports.HealthConnectPermission = exports.HealthConnectPlugin = exports.PluginRegistry = exports.HealthSyncSDK = exports.SDKEvent = exports.DEFAULT_SYNC_OPTIONS = exports.DEFAULT_SDK_CONFIG = exports.NetworkError = exports.SyncError = exports.CacheError = exports.PluginError = exports.ValidationError = exports.RateLimitError = exports.DataFetchError = exports.ConnectionError = exports.AuthenticationError = exports.ConfigurationError = exports.SDKError = exports.CacheLayer = exports.LogLevel = exports.BasePlugin = exports.ErrorAction = exports.ConnectionStatus = exports.isValidHealthData = exports.isDataType = exports.ActivityType = exports.SleepStage = exports.HeartRateContext = exports.DataQuality = exports.DataType = exports.HealthSource = void 0;
// ============================================================================
// Unified Data Models
// ============================================================================
var unified_data_1 = require("./models/unified-data");
// Enums
Object.defineProperty(exports, "HealthSource", { enumerable: true, get: function () { return unified_data_1.HealthSource; } });
Object.defineProperty(exports, "DataType", { enumerable: true, get: function () { return unified_data_1.DataType; } });
Object.defineProperty(exports, "DataQuality", { enumerable: true, get: function () { return unified_data_1.DataQuality; } });
Object.defineProperty(exports, "HeartRateContext", { enumerable: true, get: function () { return unified_data_1.HeartRateContext; } });
Object.defineProperty(exports, "SleepStage", { enumerable: true, get: function () { return unified_data_1.SleepStage; } });
Object.defineProperty(exports, "ActivityType", { enumerable: true, get: function () { return unified_data_1.ActivityType; } });
// Type Guards & Utilities
Object.defineProperty(exports, "isDataType", { enumerable: true, get: function () { return unified_data_1.isDataType; } });
Object.defineProperty(exports, "isValidHealthData", { enumerable: true, get: function () { return unified_data_1.isValidHealthData; } });
// ============================================================================
// Plugin System
// ============================================================================
var plugin_interface_1 = require("./plugins/plugin-interface");
// Enums
Object.defineProperty(exports, "ConnectionStatus", { enumerable: true, get: function () { return plugin_interface_1.ConnectionStatus; } });
Object.defineProperty(exports, "ErrorAction", { enumerable: true, get: function () { return plugin_interface_1.ErrorAction; } });
// Base Class
Object.defineProperty(exports, "BasePlugin", { enumerable: true, get: function () { return plugin_interface_1.BasePlugin; } });
// ============================================================================
// Configuration & Types
// ============================================================================
var config_1 = require("./types/config");
// Enums
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return config_1.LogLevel; } });
Object.defineProperty(exports, "CacheLayer", { enumerable: true, get: function () { return config_1.CacheLayer; } });
// Error Classes
Object.defineProperty(exports, "SDKError", { enumerable: true, get: function () { return config_1.SDKError; } });
Object.defineProperty(exports, "ConfigurationError", { enumerable: true, get: function () { return config_1.ConfigurationError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return config_1.AuthenticationError; } });
Object.defineProperty(exports, "ConnectionError", { enumerable: true, get: function () { return config_1.ConnectionError; } });
Object.defineProperty(exports, "DataFetchError", { enumerable: true, get: function () { return config_1.DataFetchError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return config_1.RateLimitError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return config_1.ValidationError; } });
Object.defineProperty(exports, "PluginError", { enumerable: true, get: function () { return config_1.PluginError; } });
Object.defineProperty(exports, "CacheError", { enumerable: true, get: function () { return config_1.CacheError; } });
Object.defineProperty(exports, "SyncError", { enumerable: true, get: function () { return config_1.SyncError; } });
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return config_1.NetworkError; } });
// Default Configurations
Object.defineProperty(exports, "DEFAULT_SDK_CONFIG", { enumerable: true, get: function () { return config_1.DEFAULT_SDK_CONFIG; } });
Object.defineProperty(exports, "DEFAULT_SYNC_OPTIONS", { enumerable: true, get: function () { return config_1.DEFAULT_SYNC_OPTIONS; } });
// ============================================================================
// Query & Response Types
// ============================================================================
var query_1 = require("./types/query");
// Enums
Object.defineProperty(exports, "SDKEvent", { enumerable: true, get: function () { return query_1.SDKEvent; } });
// ============================================================================
// Main SDK
// ============================================================================
var sdk_1 = require("./sdk");
Object.defineProperty(exports, "HealthSyncSDK", { enumerable: true, get: function () { return sdk_1.HealthSyncSDK; } });
// ============================================================================
// Plugin Registry
// ============================================================================
var plugin_registry_1 = require("./plugins/plugin-registry");
Object.defineProperty(exports, "PluginRegistry", { enumerable: true, get: function () { return plugin_registry_1.PluginRegistry; } });
// ============================================================================
// Health Connect Plugin
// ============================================================================
var health_connect_1 = require("./plugins/health-connect");
Object.defineProperty(exports, "HealthConnectPlugin", { enumerable: true, get: function () { return health_connect_1.HealthConnectPlugin; } });
Object.defineProperty(exports, "HealthConnectPermission", { enumerable: true, get: function () { return health_connect_1.HealthConnectPermission; } });
Object.defineProperty(exports, "HealthConnectRecordType", { enumerable: true, get: function () { return health_connect_1.HealthConnectRecordType; } });
Object.defineProperty(exports, "HealthConnectSleepStage", { enumerable: true, get: function () { return health_connect_1.HealthConnectSleepStage; } });
Object.defineProperty(exports, "HealthConnectExerciseType", { enumerable: true, get: function () { return health_connect_1.HealthConnectExerciseType; } });
Object.defineProperty(exports, "HealthConnectAvailability", { enumerable: true, get: function () { return health_connect_1.HealthConnectAvailability; } });
Object.defineProperty(exports, "HEALTH_CONNECT_TYPE_MAP", { enumerable: true, get: function () { return health_connect_1.HEALTH_CONNECT_TYPE_MAP; } });
Object.defineProperty(exports, "DEFAULT_HEALTH_CONNECT_CONFIG", { enumerable: true, get: function () { return health_connect_1.DEFAULT_HEALTH_CONNECT_CONFIG; } });
// ============================================================================
// Data Normalizer
// ============================================================================
var data_normalizer_1 = require("./normalizer/data-normalizer");
Object.defineProperty(exports, "DataNormalizer", { enumerable: true, get: function () { return data_normalizer_1.DataNormalizer; } });
var unit_converter_1 = require("./normalizer/unit-converter");
Object.defineProperty(exports, "UnitConverter", { enumerable: true, get: function () { return unit_converter_1.UnitConverter; } });
var validator_1 = require("./normalizer/validator");
Object.defineProperty(exports, "DataValidator", { enumerable: true, get: function () { return validator_1.DataValidator; } });
var quality_scorer_1 = require("./normalizer/quality-scorer");
Object.defineProperty(exports, "QualityScorer", { enumerable: true, get: function () { return quality_scorer_1.QualityScorer; } });
// ============================================================================
// Cache System
// ============================================================================
var cache_provider_1 = require("./cache/cache-provider");
Object.defineProperty(exports, "generateCacheKey", { enumerable: true, get: function () { return cache_provider_1.generateCacheKey; } });
Object.defineProperty(exports, "parseCacheKey", { enumerable: true, get: function () { return cache_provider_1.parseCacheKey; } });
var memory_cache_1 = require("./cache/memory-cache");
Object.defineProperty(exports, "MemoryCacheProvider", { enumerable: true, get: function () { return memory_cache_1.MemoryCacheProvider; } });
var cache_manager_1 = require("./cache/cache-manager");
Object.defineProperty(exports, "CacheManager", { enumerable: true, get: function () { return cache_manager_1.CacheManager; } });
// ============================================================================
// Utilities
// ============================================================================
var event_emitter_1 = require("./utils/event-emitter");
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return event_emitter_1.EventEmitter; } });
// ============================================================================
// Version
// ============================================================================
/**
 * Current SDK version
 */
exports.VERSION = '1.0.0';
/**
 * SDK name
 */
exports.SDK_NAME = 'HealthSync SDK';
/**
 * SDK description
 */
exports.SDK_DESCRIPTION = 'Universal health data integration platform';
