// @healthsync/core/dist/types/config.js
"use strict";
/**
 * Configuration Types
 *
 * This module defines all configuration interfaces and types for the HealthSync SDK,
 * including SDK initialization config, sync options, and error types.
 *
 * @module types/config
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SYNC_OPTIONS = exports.DEFAULT_SDK_CONFIG = exports.NetworkError = exports.SyncError = exports.CacheError = exports.PluginError = exports.ValidationError = exports.RateLimitError = exports.DataFetchError = exports.ConnectionError = exports.AuthenticationError = exports.ConfigurationError = exports.SDKError = exports.CacheLayer = exports.LogLevel = void 0;
const unified_data_1 = require("../models/unified-data");
/**
 * Log level enumeration
 *
 * @enum {string}
 */
var LogLevel;
(function (LogLevel) {
    /** No logging */
    LogLevel["NONE"] = "none";
    /** Log errors only */
    LogLevel["ERROR"] = "error";
    /** Log warnings and errors */
    LogLevel["WARN"] = "warn";
    /** Log info, warnings, and errors */
    LogLevel["INFO"] = "info";
    /** Log everything including debug messages */
    LogLevel["DEBUG"] = "debug";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
/**
 * Cache storage layer
 *
 * @enum {string}
 */
var CacheLayer;
(function (CacheLayer) {
    /** In-memory cache (fastest, session-only) */
    CacheLayer["MEMORY"] = "memory";
    /** Local database cache (SQLite, IndexedDB) */
    CacheLayer["LOCAL_DB"] = "local_db";
    /** Cloud storage cache (user's backend) */
    CacheLayer["CLOUD"] = "cloud";
    /** All cache layers */
    CacheLayer["ALL"] = "all";
})(CacheLayer || (exports.CacheLayer = CacheLayer = {}));
// ============================================================================
// Error Types
// ============================================================================
/**
 * Base error class for all SDK errors
 *
 * @class SDKError
 * @extends {Error}
 */
class SDKError extends Error {
    constructor(message, code, statusCode, details, originalError) {
        super(message);
        /** Error code */
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** HTTP status code (if applicable) */
        Object.defineProperty(this, "statusCode", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Additional error details */
        Object.defineProperty(this, "details", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Original error (if wrapped) */
        Object.defineProperty(this, "originalError", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'SDKError';
        this.code = code;
        if (statusCode !== undefined) {
            this.statusCode = statusCode;
        }
        if (details !== undefined) {
            this.details = details;
        }
        if (originalError !== undefined) {
            this.originalError = originalError;
        }
        // Maintain proper stack trace
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
}
exports.SDKError = SDKError;
/**
 * Configuration error
 *
 * @class ConfigurationError
 * @extends {SDKError}
 */
class ConfigurationError extends SDKError {
    constructor(message, details) {
        super(message, 'CONFIGURATION_ERROR', undefined, details);
        this.name = 'ConfigurationError';
    }
}
exports.ConfigurationError = ConfigurationError;
/**
 * Authentication error
 *
 * @class AuthenticationError
 * @extends {SDKError}
 */
class AuthenticationError extends SDKError {
    constructor(message, statusCode, details) {
        super(message, 'AUTHENTICATION_ERROR', statusCode, details);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * Connection error
 *
 * @class ConnectionError
 * @extends {SDKError}
 */
class ConnectionError extends SDKError {
    constructor(message, source, statusCode, details) {
        super(message, 'CONNECTION_ERROR', statusCode, details);
        /** Source that failed to connect */
        Object.defineProperty(this, "source", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'ConnectionError';
        this.source = source;
    }
}
exports.ConnectionError = ConnectionError;
/**
 * Data fetch error
 *
 * @class DataFetchError
 * @extends {SDKError}
 */
class DataFetchError extends SDKError {
    constructor(message, source, dataType, statusCode, details) {
        super(message, 'DATA_FETCH_ERROR', statusCode, details);
        /** Source where fetch failed */
        Object.defineProperty(this, "source", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Data type that failed to fetch */
        Object.defineProperty(this, "dataType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'DataFetchError';
        this.source = source;
        this.dataType = dataType;
    }
}
exports.DataFetchError = DataFetchError;
/**
 * Rate limit error
 *
 * @class RateLimitError
 * @extends {SDKError}
 */
class RateLimitError extends SDKError {
    constructor(message, resetAt, retryAfter) {
        super(message, 'RATE_LIMIT_ERROR', 429, { resetAt, retryAfter });
        /** Timestamp when rate limit resets (ISO 8601) */
        Object.defineProperty(this, "resetAt", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Retry after duration in seconds */
        Object.defineProperty(this, "retryAfter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'RateLimitError';
        this.resetAt = resetAt;
        this.retryAfter = retryAfter;
    }
}
exports.RateLimitError = RateLimitError;
/**
 * Validation error
 *
 * @class ValidationError
 * @extends {SDKError}
 */
class ValidationError extends SDKError {
    constructor(message, fields, details) {
        super(message, 'VALIDATION_ERROR', 400, { ...details, fields });
        /** Fields that failed validation */
        Object.defineProperty(this, "fields", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'ValidationError';
        this.fields = fields;
    }
}
exports.ValidationError = ValidationError;
/**
 * Plugin error
 *
 * @class PluginError
 * @extends {SDKError}
 */
class PluginError extends SDKError {
    constructor(message, pluginId, statusCode, details) {
        super(message, 'PLUGIN_ERROR', statusCode, { ...details, pluginId });
        /** Plugin ID where error occurred */
        Object.defineProperty(this, "pluginId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'PluginError';
        this.pluginId = pluginId;
    }
}
exports.PluginError = PluginError;
/**
 * Cache error
 *
 * @class CacheError
 * @extends {SDKError}
 */
class CacheError extends SDKError {
    constructor(message, layer, details) {
        super(message, 'CACHE_ERROR', undefined, { ...details, layer });
        /** Cache layer where error occurred */
        Object.defineProperty(this, "layer", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'CacheError';
        this.layer = layer;
    }
}
exports.CacheError = CacheError;
/**
 * Sync error
 *
 * @class SyncError
 * @extends {SDKError}
 */
class SyncError extends SDKError {
    constructor(message, phase, source, dataType, details) {
        super(message, 'SYNC_ERROR', undefined, { ...details, phase, source, dataType });
        /** Source where sync failed */
        Object.defineProperty(this, "source", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Data type where sync failed */
        Object.defineProperty(this, "dataType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /** Phase where sync failed */
        Object.defineProperty(this, "phase", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'SyncError';
        if (source !== undefined) {
            this.source = source;
        }
        if (dataType !== undefined) {
            this.dataType = dataType;
        }
        this.phase = phase;
    }
}
exports.SyncError = SyncError;
/**
 * Network error
 *
 * @class NetworkError
 * @extends {SDKError}
 */
class NetworkError extends SDKError {
    constructor(message, statusCode, isRetryable = true, details) {
        super(message, 'NETWORK_ERROR', statusCode, { ...details, isRetryable });
        /** Whether the error is retryable */
        Object.defineProperty(this, "isRetryable", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = 'NetworkError';
        this.isRetryable = isRetryable;
    }
}
exports.NetworkError = NetworkError;
/**
 * Default SDK configuration values
 */
exports.DEFAULT_SDK_CONFIG = {
    environment: 'production',
    cache: {
        enabled: true,
        layers: [CacheLayer.MEMORY, CacheLayer.LOCAL_DB],
        defaultTTL: 24 * 60 * 60 * 1000, // 24 hours
        maxSize: 50, // 50 MB
        useLRU: true,
    },
    retry: {
        maxAttempts: 3,
        initialDelay: 1000,
        maxDelay: 30000,
        backoffMultiplier: 2,
        useJitter: true,
        timeout: 30000,
    },
    logLevel: LogLevel.WARN,
    plugins: [],
    defaultDataTypes: [unified_data_1.DataType.STEPS, unified_data_1.DataType.HEART_RATE, unified_data_1.DataType.SLEEP, unified_data_1.DataType.ACTIVITY],
    autoSync: false,
    syncInterval: 0,
    maxDataAge: 90, // 90 days
    enableRealtimeUpdates: false,
};
/**
 * Default sync options
 */
exports.DEFAULT_SYNC_OPTIONS = {
    forceFull: false,
    background: false,
    maxConcurrency: 3,
    priority: 1,
};
