// @healthsync/core/dist/esm/types/query.js
/**
 * Query and Response Types
 *
 * This module defines the types for querying health data and handling responses.
 *
 * @module types/query
 */
/**
 * Event types emitted by the SDK
 *
 * @enum {string}
 */
export var SDKEvent;
(function (SDKEvent) {
    /** SDK initialized */
    SDKEvent["INITIALIZED"] = "initialized";
    /** Plugin registered */
    SDKEvent["PLUGIN_REGISTERED"] = "plugin_registered";
    /** Plugin removed */
    SDKEvent["PLUGIN_REMOVED"] = "plugin_removed";
    /** Connection established */
    SDKEvent["CONNECTED"] = "connected";
    /** Connection lost */
    SDKEvent["DISCONNECTED"] = "disconnected";
    /** Data update received */
    SDKEvent["DATA_UPDATE"] = "data_update";
    /** Sync started */
    SDKEvent["SYNC_STARTED"] = "sync_started";
    /** Sync completed */
    SDKEvent["SYNC_COMPLETED"] = "sync_completed";
    /** Sync failed */
    SDKEvent["SYNC_FAILED"] = "sync_failed";
    /** Error occurred */
    SDKEvent["ERROR"] = "error";
})(SDKEvent || (SDKEvent = {}));
