// @healthsync/core/dist/plugins/health-connect/index.js
"use strict";
/**
 * Health Connect Plugin
 *
 * Android Health Connect integration for unified health and fitness data access.
 *
 * @module plugins/health-connect
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_HEALTH_CONNECT_CONFIG = exports.HEALTH_CONNECT_TYPE_MAP = exports.HealthConnectAvailability = exports.HealthConnectExerciseType = exports.HealthConnectSleepStage = exports.HealthConnectRecordType = exports.HealthConnectPermission = exports.HealthConnectPlugin = void 0;
var health_connect_plugin_1 = require("./health-connect-plugin");
Object.defineProperty(exports, "HealthConnectPlugin", { enumerable: true, get: function () { return health_connect_plugin_1.HealthConnectPlugin; } });
var types_1 = require("./types");
Object.defineProperty(exports, "HealthConnectPermission", { enumerable: true, get: function () { return types_1.HealthConnectPermission; } });
Object.defineProperty(exports, "HealthConnectRecordType", { enumerable: true, get: function () { return types_1.HealthConnectRecordType; } });
Object.defineProperty(exports, "HealthConnectSleepStage", { enumerable: true, get: function () { return types_1.HealthConnectSleepStage; } });
Object.defineProperty(exports, "HealthConnectExerciseType", { enumerable: true, get: function () { return types_1.HealthConnectExerciseType; } });
Object.defineProperty(exports, "HealthConnectAvailability", { enumerable: true, get: function () { return types_1.HealthConnectAvailability; } });
Object.defineProperty(exports, "HEALTH_CONNECT_TYPE_MAP", { enumerable: true, get: function () { return types_1.HEALTH_CONNECT_TYPE_MAP; } });
Object.defineProperty(exports, "DEFAULT_HEALTH_CONNECT_CONFIG", { enumerable: true, get: function () { return types_1.DEFAULT_HEALTH_CONNECT_CONFIG; } });
